src/ol/style/expressions.js

```diff
@@ -366,40 +423,81 @@ function assertArgsMaxCount(args, count) {
 function assertArgsEven(args) {
   if (args.length % 2 !== 0) {
     throw new Error(
-      `An even amount of arguments was expected, got ${args} instead`
+      `An even amount of arguments was expected, got ${JSON.stringify(
+        args
+      )} instead`
     );
   }
 }
 function assertArgsOdd(args) {
   if (args.length % 2 === 0) {
     throw new Error(
-      `An odd amount of arguments was expected, got ${args} instead`
+      `An odd amount of arguments was expected, got ${JSON.stringify(
+        args
+      )} instead`
     );
   }
 }
-function assertUniqueInferredType(args, types) {
-  if (!isTypeUnique(types)) {
+function assertNotEmptyType(args, types, descriptor) {
+  if (types === ValueTypes.NONE) {
     throw new Error(
-      `Could not infer only one type from the following expression: ${JSON.stringify(
+      `No matching type was found for the following expression ${descriptor}: ${JSON.stringify(
         args
       )}`
     );
   }
 }
+function assertSingleType(args, types, descriptor) {
+  assertNotEmptyType(args, types, descriptor);
+  if (!isTypeUnique(types)) {
+    throw new Error(
+      `Expected to have a unique type for the following expression ${descriptor}: ${JSON.stringify(
+        args
+      )}
+Got the following types instead: ${printTypes(types)}`
+    );
+  }
+}
+function assertOfType(args, types, expectedTypes, descriptor) {
+  if ((types & expectedTypes) === ValueTypes.NONE) {
+    throw new Error(
+      `Expected the ${descriptor} type of the following expression: ${JSON.stringify(
+        args
+      )} to be of the following types: ${printTypes(expectedTypes)}
+Got these types instead: ${printTypes(types)}`
+    );
+  }
+}
 
 Operators['get'] = {
   getReturnType: function (args) {
+    if (args.length === 2) {
+      const hint = args[1];
+      return getTypeFromHint(/** @type {string} */ (hint));
+    }
     return ValueTypes.ANY;
   },
-  toGlsl: function (context, args) {
-    assertArgsCount(args, 1);
+  toGlsl: function (context, args, expectedType) {
+    assertArgsMinCount(args, 1);
+    assertArgsMaxCount(args, 2);
     assertString(args[0]);
-    const value = args[0].toString();
-    if (!context.attributes.includes(value)) {
-      context.attributes.push(value);
+    const outputType = expectedType & Operators['get'].getReturnType(args);
+    assertSingleType(['get', ...args], outputType, '');
+    const name = args[0].toString();
+    const existing = context.attributes.find((a) => a.name === name);
+    if (!existing) {
+      context.attributes.push({
+        name: name,
+        type: outputType,
+      });
+    } else if (outputType !== existing.type) {
+      throw new Error(
+        `The following attribute was used in different places with incompatible types: ${name}
+Types were: ${printTypes(existing.type)} and ${printTypes(outputType)}`
+      );
     }
     const prefix = context.inFragmentShader ? 'v_' : 'a_';
-    return prefix + value;
+    return prefix + name;
```

Comment:
Doesn’t need to be in this branch, but we should prefix these with something that avoids conflicts with names that we might want to use (e.g. `a_user_` or similar).

Reply:
Leaving this for another PR as this one is already quite full! thanks for the suggestion, definitely makes sense